subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java

```diff
@@ -115,180 +110,101 @@ public ClassSetDiff findChangesSince(ClassSetAnalysis other) {
-                                         Set<String> privateResultClasses,
-                                         Set<String> accessibleResultClasses,
-                                         Set<GeneratedResource> resultResources,
-                                         Iterable<String> privateDependentClasses,
-                                         Iterable<String> accessibleDependentClasses) {
-
-        for (String privateDependentClass : privateDependentClasses) {
-            if (!visitedClasses.add(privateDependentClass)) {
+        while (!remaining.isEmpty()) {
+            String accessibleDependent = remaining.pop();
+            if (!visited.add(accessibleDependent)) {
                 continue;
             }
-            privateResultClasses.add(privateDependentClass);
-            DependentsSet currentDependents = getDependents(privateDependentClass);
-            if (!currentDependents.isDependencyToAll()) {
-                resultResources.addAll(currentDependents.getDependentResources());
+            accessibleDependents.add(accessibleDependent);
+            DependentsSet dependents = findDirectDependents(accessibleDependent);
+            if (dependents.isDependencyToAll()) {
+                return dependents;
             }
+            dependentResources.addAll(dependents.getDependentResources());
+            privateDependents.addAll(dependents.getPrivateDependentClasses());
+            remaining.addAll(dependents.getAccessibleDependentClasses());
         }
 
-        processTransitiveDependentClasses(visitedClasses, accessibleResultClasses, resultResources, accessibleDependentClasses);
+        privateDependents.remove(className);
+        accessibleDependents.remove(className);
+        return DependentsSet.dependents(privateDependents, accessibleDependents, dependentResources);
     }
 
-    private void processTransitiveDependentClasses(Set<String> visitedClasses,
-                                                   Set<String> accessibleResultClasses,
-                                                   Set<GeneratedResource> resultResources,
-                                                   Iterable<String> accessibleDependentClasses) {
-        Deque<String> remainingAccessibleDependentClasses = new ArrayDeque<>();
-        for (String accessibleDependentClass : accessibleDependentClasses) {
-            remainingAccessibleDependentClasses.add(accessibleDependentClass);
-        }
-
-        while (!remainingAccessibleDependentClasses.isEmpty()) {
-            String accessibleDependentClass = remainingAccessibleDependentClasses.pop();
-            if (!visitedClasses.add(accessibleDependentClass)) {
-                continue;
-            }
-            accessibleResultClasses.add(accessibleDependentClass);
-            DependentsSet currentDependents = getDependents(accessibleDependentClass);
-            if (!currentDependents.isDependencyToAll()) {
```

Comment:
This was swallowing dependencies to all. All the "processConstantDependents" code futher down was a workaround trying to get those dependents back. None of it is necessary anymore now that the algorithm returns dependencyToAll when one is found.